test: add performance tests for the market details page

chrisleewilcox opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->
This commit introduces a new test suite for verifying the functionality of market details tabs in the predictions feature. The tests cover navigation to the Trade tab, visibility of the market details screen, and the correct loading of content in the About and Outcomes tabs. Assertions ensure that the expected information is displayed correctly, enhancing the reliability of the predictions feature.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MMQA-1077

## **Manual testing steps**

```gherkin
Feature: Predictions Market Details
  @smoke @4g
  Scenario Outline: Measure different market categories load performance
    Given the user is on the Wallet screen
    And the user has opened the Predict Market List
    When the user switches to the "<category>" tab
    And the user taps on market index <market_index>
    Then the Market Details screen should load within acceptable time
    And the About tab should load within acceptable time
    And the Outcomes tab should load within acceptable time

    Examples:
      | category  | market_index |
      | Trending  | 1            |

```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->
browserstack test run: https://app-automate.browserstack.com/builds/d6ca8047a88fea3b225a45c2a20a74a7eefb4acb/sessions/eec871550c495cefaaea05ecfeccd9f69c18ee22?auth_token=870f686e6fb0fc1ab2847aa0fb921f8e65ee91561c3b086c502209c71f0b723e

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a performance test for Predict Market Details load times and introduces supporting screen objects, selectors, and a predict action in the wallet modal.
> 
> - **Tests (performance)**:
>   - Add `appwright/tests/performance/predict/predict-market-details.spec.js` to measure load times for opening Predict, market list, market details, and About/Outcomes tabs; records metrics via `performanceTracker`.
> - **Screen Objects**:
>   - Add `wdio/screen-objects/PredictMarketListScreen.js` with device-aware selectors and actions (`isContainerDisplayed`, `tapMarketCard`, `tapCategoryTab`).
>   - Add `wdio/screen-objects/PredictDetailsScreen.js` with visibility checks and tab actions (`tapAboutTab`, `tapOutcomesTab`, content assertions).
> - **Modal Actions**:
>   - Update `wdio/screen-objects/Modals/WalletActionModal.js` to include `predictButton` and `tapPredictButton`.
> - **Selectors**:
>   - Extend `e2e/selectors/Predict/Predict.selectors.ts` with Predict Market Details tab label IDs (`POSITIONS_TAB_LABEL`, `OUTCOMES_TAB_LABEL`, `ABOUT_TAB_LABEL`) and related IDs used by new tests.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit da66a0de1e9305a446e0879b9fc5cdad9fbb3058. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

chrisleewilcox committed:
test: add end-to-end tests for market details in predictions

This commit introduces a new test suite for verifying the functionality of market details tabs in the predictions feature. The tests cover navigation to the Trade tab, visibility of the market details screen, and the correct loading of content in the About and Outcomes tabs. Assertions ensure that the expected information is displayed correctly, enhancing the reliability of the predictions feature.

chrisleewilcox committed:
Revert "test: add end-to-end tests for market details in predictions"

This reverts commit 0916763b56689bd7d25ab364fefc27a169c3adb6.

chrisleewilcox committed:
test: add performance tests for market details in predictions

This commit introduces new performance tests for the market details page, measuring load times for various interactions under 4G network conditions. The tests track the time taken to navigate to the predictions tab, load the market list, and display details for both the About and Outcomes tabs. Additionally, new configurations for Android and iOS BrowserStack projects have been added to the appwright configuration, enhancing the testing framework.

chrisleewilcox committed:
Merge branch 'main' into test-predict-market-details

chrisleewilcox committed:
refactor: update performance test timers and selectors for market details

- Removed unnecessary timer for actions modal visibility in the performance test.
- Updated the performance test to correctly aggregate timer durations.
- Refactored `PredictDetailsScreen` to use new selectors for tab labels.

chrisleewilcox committed:
Merge branch 'main' into test-predict-market-details

chrisleewilcox committed:
fix(selectors): reorder tab labels in PredictMarketDetailsSelectorsIDs for consistency

- Updated the order of tab labels in `PredictMarketDetailsSelectorsIDs` to ensure consistency with their respective tabs.
- Adjusted the `tap` method in `PredictDetailsScreen` to await the resolved reference of `aboutTab` for improved reliability.

javiergarciavera commented on appwright/appwright.config.ts:
```diff
@@ -101,5 +101,33 @@ export default defineConfig({
         expectTimeout: 30 * 1000, //90 seconds  increased since login the app takes longer
       },
     },
+    {
```
Do we need to have a different "project" in the config just for prediction tests?
I mean, we only differenciatte between login/onboarding flows because they use different builds (one with an SRP built in)

javiergarciavera commented on package.json:
```diff
@@ -106,9 +106,11 @@
     "test:reassure:baseline": "REASSURE=true yarn reassure --baseline",
     "test:reassure:branch": "REASSURE=true yarn reassure --branch && node -e \"const r=require('./.reassure/output.json'); if ((r.significant||[]).length>0) { console.error('Reassure: significant regressions detected'); process.exit(1);} else { process.exit(0);} \"",
     "run-appwright:android-bs": "yarn appwright test --project browserstack-android --config appwright/appwright.config.ts",
+    "run-appwright:android-predict-bs": "yarn appwright test --project android-predict-bs --config appwright/appwright.config.ts",
```
same here, unless we need to run them indepentenly, I do not see the point

javiergarciavera commented on package.json:
```diff
@@ -106,9 +106,11 @@
     "run-appwright:android-onboarding-bs": "yarn appwright test --project android-onboarding --config appwright/appwright.config.ts",
     "run-appwright:ios-onboarding-bs": "yarn appwright test --project ios-onboarding --config appwright/appwright.config.ts",
     "run-appwright:ios-bs": "yarn appwright test --project browserstack-ios --config appwright/appwright.config.ts",
+    "run-appwright:ios-predict-bs": "yarn appwright test --project ios-predict-bs --config appwright/appwright.config.ts",
```
same

chrisleewilcox replied:
removed.

chrisleewilcox committed:
refactor(appwright): remove unused Android and iOS predict test configurations

- Deleted the `android-predict-bs` and `ios-predict-bs` test configurations from `package.json` and `appwright.config.ts` as they are no longer needed.
- This cleanup helps streamline the test setup and reduces potential confusion regarding obsolete test projects.

chrisleewilcox committed:
fix(tests): update total time calculation in predict-market-details test

- Refactored the total time calculation in the `predict-market-details.spec.js` test to use the `getDuration()` method for improved accuracy.
- This change ensures that the total load time is calculated correctly, enhancing the reliability of performance testing.

chrisleewilcox committed:
refactor(selectors): remove unused getElementByTextContains method

- Deleted the `getElementByTextContains` method from `AppwrightSelectors` as it was no longer needed, streamlining the selector functionality.
- This change helps reduce code complexity and improve maintainability.

chrisleewilcox commented on e2e/framework/AppwrightSelectors.ts:
```diff
@@ -22,6 +22,23 @@ export default class AppwrightSelectors {
+    if (isAndroid) {
+      // Android: text or content-desc contains
+      const xpath = `//*[contains(@text,'${text}') or contains(@content-desc,'${text}')]`;
+      return await AppwrightSelectors.getElementByXpath(testDevice, xpath);
```
Yeah it's basically the same.  Removed.

chrisleewilcox committed:
refactor(tests): update performance logging in predict-market-details test

- Refactored the logging statements in the `predict-market-details.spec.js` test to utilize the `getDuration()` method for improved clarity and consistency in performance metrics.
- This change enhances the readability of the performance output during test execution.

hjetpoluru reviewed (approved):
LGTM
